### PR TITLE
Allow DH Parameters at any position

### LINKED ISSUE #17326: SSL_CONF_cmd(.): DH Parameters must be first in file, since OpenSSL 3.0
With OpenSSL 1.0.2, the [CONF](https://www.openssl.org/docs/man3.0/man3/SSL_CONF_cmd.html) API was added. With that, an end-user is able to configure things directly within the underlying OpenSSL. In other words, the app itself does not need to offer a specific configuration switch but just offer access to the underlying OpenSSL via its CONF API. This is useful to configure things like DH parameters. With that, the user gives a path to a PEM file. In that file, till OpenSSL 1.1.1, the DH parameters could be at the first, last, or any position. Since OpenSSL 3.0, the DH parameters must come first in that file.

The blaming commit is 163f6dc. It changed from the now deprecated `PEM_read_bio_DHparams(.)` to the new `OSSL_DECODER_from_bio(.)`. However, the latter does not filter via the supplied decoder ‘keytype’, which is `DH` in this case. Instead, `OSSL_DECODER_from_bio(.)` → `decoder_process(.)` → `pem2der_decode(.)` → `read_pem(.)` → `PEM_read_bio(.)` goes for the first header in the file (at the current file position), for example, a private key or a server certificate.

Of the apps,
```shell
openssl s_server -nocert -dhparam "./server.pem" -cipher aNULL+kDHE+AES@SECLEVEL=0
```
does it correctly as it loops, see
⤷ load_keyparams[_suppress]
⤷ load_key_certs_crls_suppress

On GitHub, several issues are related to that non-filtering. However, if I not misread those reports, that non-filtering is intended. Anyway, would be great to get a confirmation. Then, I create a Pull Request which calls `OSSL_DECODER_from_bio(.)` again and again until (a) the parameters are found or (b) end-of-file is reached.

By the way, I do not understand yet, why the app `s_server` does not use the CONF API for DH parameters as well. For other command-line arguments, `s_server` uses that API. This might have revealed this issue here earlier.

By the way, by the way, I have to [shoot into feet](https://github.com/openssl/openssl/pull/13138#issuecomment-709449970) because I have an embedded device which is so slow, I have to go for less security bits in DH [compared](https://github.com/openssl/openssl/commit/7646610) to the bits of my private key.

Below are two minimum examples, one using the CONF API:
```c
#include <stdio.h>
#include <openssl/ssl.h>

int main(void) {
	SSL_CTX *ctx = SSL_CTX_new(TLS_client_method());
	SSL_CONF_CTX *cctx = SSL_CONF_CTX_new();

	SSL_CONF_CTX_set_ssl_ctx(cctx, ctx);
	SSL_CONF_CTX_set_flags(cctx, SSL_CONF_FLAG_CERTIFICATE
		| SSL_CONF_FLAG_SERVER
		| SSL_CONF_FLAG_FILE);

	if (2 == SSL_CONF_cmd(cctx, "DHParameters", "server.pem")) {
		puts("Success");
	}

	return 0;
}
```
and the next one is an minimum extract of how `SSL_CONF_cmd(.)` uses the SSL API:
```c
#include <stdio.h>                 /* for NULL, puts */
#include <openssl/bio.h>           /* for BIO_free, BIO_new_file */
#include <openssl/core_dispatch.h> /* for OSSL_KEYMGMT_SELECT_DOMAIN_P... */
#include <openssl/decoder.h>       /* for OSSL_DECODER_CTX_free, ... */
#include <openssl/evp.h>           /* for EVP_PKEY_free */
#include <openssl/types.h>         /* for BIO, EVP_PKEY, OSSL_DECODER_CTX */

int main(void) {
	EVP_PKEY *dhpkey;
	OSSL_DECODER_CTX *decoderctx
		= OSSL_DECODER_CTX_new_for_pkey(&dhpkey, "PEM", NULL, "DH",
			OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS, NULL, NULL);
	BIO *bio = BIO_new_file("server.pem", "r");
	if (OSSL_DECODER_from_bio(decoderctx, bio)) {
		//SSL_CTX_set0_tmp_dh_pkey(..., dhpkey);
		puts("Success");
	}
	EVP_PKEY_free(dhpkey);
	OSSL_DECODER_CTX_free(decoderctx);
	BIO_free(bio);
	return 0;
}
```
Tested with branch 3.0.1 and the branch master as of today.

### PR DESCRIPTION
**Pull request by traud**
Fixes #17326

### CODE REVIEW

**Comment by traud** [comment]
Mhm. While going through the source code if other `OSSL_DECODER_from_bio(.)` usages are affected as well, I found [pem_pkey](https://github.com/openssl/openssl/blob/66a7c9f34b46edd462d647ae2febe8276bb9b4f7/crypto/pem/pem_pkey.c#L61): `OSSL_DECODER_from_bio(.)` might not advance the file position, I think. At least that other loop is coded that way. I am planning to improve my Pull Request to align both code places / loops. Can someone put this Pull Request in work-as-progress here?

File: ssl/ssl_conf.c [hunk 1]
```diff
@@ -597,10 +597,13 @@ static int cmd_DHParameters(SSL_CONF_CTX *cctx, const char *value)
...
-                || !OSSL_DECODER_from_bio(decoderctx, in)) {
-            OSSL_DECODER_CTX_free(decoderctx);
+        if (decoderctx == NULL)
             goto end;
+
+        while (!OSSL_DECODER_from_bio(decoderctx, in)) {
+            if (BIO_eof(in)) {
+                OSSL_DECODER_CTX_free(decoderctx);
+                goto end;
         }
```
Comment (paulidale) [inline ssl/ssl_conf.c:607]: Needs a closing `}` here.

**Comment by kroeckx** [comment]
Can you add a test?

**Comment by t8m** [comment]
We have alternative non-stalled PR  #18041. Closing this.

**Pull request closed without merging**